lib/domain.js

```diff
@@ -48,46 +48,39 @@ Object.defineProperty(process, 'domain', {
   }
 });
 
-let asyncHook = null;
 const pairing = new Map();
-function enableHook() {
-
-  if (asyncHook !== null) {
-    return;
-  }
-
-  asyncHook = async_hooks.createHook({
-    init(asyncId, type, triggerAsyncId, resource) {
-
-      if (process.domain !== null) { // if this operation is created while in a domain, let's mark it
-        pairing.set(asyncId, process.domain);
-        resource.domain = process.domain;
-        if (resource.promise !== undefined && resource.promise instanceof Promise) {
-          // resource.promise instanceof Promise make sure that the promise comes from the same context
-          // see https://github.com/nodejs/node/issues/15673
-          resource.promise.domain = process.domain;
-        }
+const asyncHook = async_hooks.createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+
+    if (process.domain !== null && process.domain !== undefined) { // if this operation is created while in a domain, let's mark it
+      pairing.set(asyncId, process.domain);
+      resource.domain = process.domain;
+      if (resource.promise !== undefined && resource.promise instanceof Promise) {
+        // resource.promise instanceof Promise make sure that the promise comes from the same context
+        // see https://github.com/nodejs/node/issues/15673
+        resource.promise.domain = process.domain;
       }
-    },
-    before(asyncId) {
-
-      if (pairing.has(asyncId)) { // enter domain for this cb
-        process.domain = pairing.get(asyncId);
-      }
-    },
-    after(asyncId) {
+    }
+  },
+  before(asyncId) {
 
-      if (pairing.has(asyncId)) { // exit domain for this cb
-        process.domain = null;
-      }
-    },
-    destroy(asyncId) {
+    if (pairing.has(asyncId)) { // enter domain for this cb
+      pairing.get(asyncId).enter();
+    }
+  },
+  after(asyncId) {
 
-      pairing.delete(asyncId); // cleaning up
+    if (pairing.has(asyncId)) { // exit domain for this cb
+      pairing.get(asyncId).exit();
     }
-  });
-  asyncHook.enable();
-}
+  },
+  destroy(asyncId) {
+
+    pairing.delete(asyncId); // cleaning up
+  }
+});
+asyncHook.enable();
```

Comment:
Move this to the `Domain` constructor. If the `asyncHook` is already enabled it will just be a noop. Hooks can only be enabled once.